File: Database_Integration/java/ImportFundsXml.java
// =============================================================================
// IMPORT  —  FundsXML file  ->  relational database  (Java, SQLite via JDBC).
//
// Standalone, copy-me example of ONE direction (FundsXML -> DB). The reverse
// is a separate program, ExportFundsXml.java. Over-commented as documentation.
//
// DB SCHEMA  ../ddl/schema.sql  (document -> fund -> portfolio -> position;
//   share_class per fund; asset document-scoped). Creates the schema in a
//   fresh SQLite DB, then loads the file.
//
// RUN  (sqlite-jdbc fetched by tools/fetch-tools.sh into .lib/)
//   tools/fetch-tools.sh
//   CP=.lib/sqlite-jdbc-3.46.1.3.jar
//   javac -cp "$CP" -d /tmp/db Database_Integration/java/ImportFundsXml.java
//   java --enable-native-access=ALL-UNNAMED -cp "$CP:/tmp/db" \
//     ImportFundsXml fx.db FundsXML_Files/4.2.9/positions/Multi-Fund_Positions.xml
//
// DEPENDENCIES  org.xerial:sqlite-jdbc + the JDK. XML via native javax.xml
//   DOM/XPath — NO JAXB (a thin DOM binding is smaller and version-tolerant).
//
// FUNDSXML ASSUMPTIONS
//   * No XML namespace -> bare element names in XPath.
//   * Many <Fund>/<Portfolio>/<Position>: all iterated; 1-based *_seq columns
//     preserve order so the separate export reproduces the original document.
//   * Positions link to AssetMasterData by shared <UniqueID> -> `asset` is
//     document-scoped.
//
// SECURITY  the DOM parser forbids DOCTYPE/external entities (XXE).
// =============================================================================
import java.io.FileInputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.Statement;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathFactory;
import org.w3c.dom.Document;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;

public class ImportFundsXml {

    static final java.util.Set<String> POSITION_KINDS = java.util.Set.of(
        "Equity", "Bond", "ShareClass", "Warrant", "Certificate", "Option",
        "Future", "FXForward", "Swap", "Repo", "RealEstate", "CallMoney",
        "Account", "Generic");
    static final java.util.Map<String, String> QTY_ELEM = java.util.Map.of(
        "Equity", "Units", "Warrant", "Units", "Certificate", "Units",
        "Bond", "Nominal", "ShareClass", "Shares",
        "Option", "Contracts", "Future", "Contracts");

    static XPath xpath() { return XPathFactory.newInstance().newXPath(); }

    /** First matching descendant's text (bare names; no namespace). */
    static String t(Node ctx, String expr) throws Exception {
        Node n = (Node) xpath().evaluate(expr, ctx, XPathConstants.NODE);
        return n == null ? null : n.getTextContent();
    }
    static Double num(String s) {
        return (s == null || s.isBlank()) ? null : Double.valueOf(s.trim());
    }
    static void setNum(PreparedStatement ps, int i, Double v) throws Exception {
        if (v == null) ps.setNull(i, java.sql.Types.DECIMAL);
        else ps.setDouble(i, v);
    }

    public static void main(String[] args) throws Exception {
        if (args.length != 2) {
            System.err.println("usage: ImportFundsXml <db> <fundsxml.xml>");
            System.exit(2);
        }
        String db = args[0], xml = args[1];

        // ---- parse the FundsXML file (hardened) ---------------------------
        DocumentBuilderFactory f = DocumentBuilderFactory.newInstance();
        f.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
        f.setFeature("http://xml.org/sax/features/external-general-entities", false);
        f.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
        f.setExpandEntityReferences(false);
        Document doc;
        try (FileInputStream in = new FileInputStream(xml)) {
            doc = f.newDocumentBuilder().parse(in);
        }
        XPath xp = xpath();

        try (Connection c = DriverManager.getConnection("jdbc:sqlite:" + db)) {
            c.createStatement().execute("PRAGMA foreign_keys = ON");

            // ---- create the schema (fresh DB) -----------------------------
            Path ddl = Path.of(System.getProperty("user.dir"),
                "Database_Integration", "ddl", "schema.sql");
            if (!Files.exists(ddl)) ddl = Path.of("..", "ddl", "schema.sql");
            StringBuilder clean = new StringBuilder();
            for (String line : Files.readString(ddl).split("\n")) {
                int cut = line.indexOf("--");        // strip line comments
                clean.append(cut >= 0 ? line.substring(0, cut) : line)
                     .append('\n');
            }
            try (Statement st = c.createStatement()) {
                for (String stmt : clean.toString().split(";"))
                    if (!stmt.strip().isEmpty()) st.execute(stmt);
            }

            c.setAutoCommit(false);
            Node cd = (Node) xp.evaluate("/FundsXML4/ControlData", doc,
                XPathConstants.NODE);
            String docId = t(cd, "UniqueDocumentID");

            try (PreparedStatement ps = c.prepareStatement(
                    "INSERT INTO document VALUES (?,?,?,?,?,?,?,?,?)")) {
                ps.setString(1, docId);
                ps.setString(2, t(cd, "DocumentGenerated"));
                ps.setString(3, t(cd, "Version"));
                ps.setString(4, t(cd, "ContentDate"));
                ps.setString(5, t(cd, "DataOperation"));
                ps.setString(6, t(cd, "DataSupplier/SystemCountry"));
                ps.setString(7, t(cd, "DataSupplier/Short"));
                ps.setString(8, t(cd, "DataSupplier/Name"));
                ps.setString(9, t(cd, "DataSupplier/Type"));
                ps.executeUpdate();
            }

            NodeList funds = (NodeList) xp.evaluate("/FundsXML4/Funds/Fund",
                doc, XPathConstants.NODESET);
            for (int fi = 0; fi < funds.getLength(); fi++) {
                Node fund = funds.item(fi);
                int fundSeq = fi + 1;                // 1-based document order
                String ccy = t(fund, "Currency");
                Node tav = (Node) xp.evaluate(
                    "FundDynamicData/TotalAssetValues/TotalAssetValue",
                    fund, XPathConstants.NODE);
                try (PreparedStatement ps = c.prepareStatement(
                        "INSERT INTO fund VALUES (?,?,?,?,?,?,?,?)")) {
                    ps.setString(1, docId);
                    ps.setInt(2, fundSeq);
                    ps.setString(3, t(fund, "Identifiers/LEI"));
                    ps.setString(4, t(fund, "Names/OfficialName"));
                    ps.setString(5, ccy);
                    ps.setString(6, t(fund, "SingleFundFlag"));
                    ps.setString(7, t(tav, "NavDate"));
                    ps.setDouble(8, Double.parseDouble(t(tav,
                        "TotalNetAssetValue/Amount[@ccy='" + ccy + "']")));
                    ps.executeUpdate();
                }

                NodeList scs = (NodeList) xp.evaluate(
                    "SingleFund/ShareClasses/ShareClass", fund,
                    XPathConstants.NODESET);
                for (int si = 0; si < scs.getLength(); si++) {
                    Node sc = scs.item(si);
                    try (PreparedStatement ps = c.prepareStatement(
                            "INSERT INTO share_class VALUES (?,?,?,?,?,?,?,?)")) {
                        ps.setString(1, docId);
                        ps.setInt(2, fundSeq);
                        ps.setString(3, t(sc, "Identifiers/ISIN"));
                        ps.setString(4, t(sc, "Names/OfficialName"));
                        ps.setString(5, t(sc, "Currency"));
                        setNum(ps, 6, num(t(sc, "Prices/Price/NavPrice")));
                        setNum(ps, 7, num(t(sc, "TotalAssetValues/"
                            + "TotalAssetValue/TotalNetAssetValue/Amount[@ccy='"
                            + ccy + "']")));
                        setNum(ps, 8, num(t(sc, "TotalAssetValues/"
                            + "TotalAssetValue/SharesOutstanding")));
                        ps.executeUpdate();
                    }
                }

                NodeList ports = (NodeList) xp.evaluate(
                    "FundDynamicData/Portfolios/Portfolio", fund,
                    XPathConstants.NODESET);
                for (int pi = 0; pi < ports.getLength(); pi++) {
                    Node port = ports.item(pi);
                    int portSeq = pi + 1;
                    try (PreparedStatement ps = c.prepareStatement(
                            "INSERT INTO portfolio VALUES (?,?,?,?)")) {
                        ps.setString(1, docId);
                        ps.setInt(2, fundSeq);
                        ps.setInt(3, portSeq);
                        ps.setString(4, t(port, "NavDate"));
                        ps.executeUpdate();
                    }
                    NodeList poss = (NodeList) xp.evaluate(
                        "Positions/Position", port, XPathConstants.NODESET);
                    for (int qi = 0; qi < poss.getLength(); qi++) {
                        Node pos = poss.item(qi);
                        String kind = null;
                        for (Node ch = pos.getFirstChild(); ch != null;
                                ch = ch.getNextSibling())
                            if (ch.getNodeType() == Node.ELEMENT_NODE
                                && POSITION_KINDS.contains(ch.getNodeName())) {
                                kind = ch.getNodeName();
                                break;
                            }
                        Double qty = (kind != null && QTY_ELEM.containsKey(kind))
                            ? num(t(pos, kind + "/" + QTY_ELEM.get(kind)))
                            : null;
                        try (PreparedStatement ps = c.prepareStatement(
                                "INSERT INTO position VALUES (?,?,?,?,?,?,?,?,?,?,?)")) {
                            ps.setString(1, docId);
                            ps.setInt(2, fundSeq);
                            ps.setInt(3, portSeq);
                            ps.setInt(4, qi + 1);
                            ps.setString(5, t(pos, "UniqueID"));
                            ps.setString(6, t(pos, "Identifiers/ISIN"));
                            ps.setString(7, t(pos, "Currency"));
                            ps.setDouble(8, Double.parseDouble(t(pos,
                                "TotalValue/Amount[@ccy='" + ccy + "']")));
                            ps.setDouble(9, Double.parseDouble(
                                t(pos, "TotalPercentage")));
                            ps.setString(10, kind);
                            setNum(ps, 11, qty);
                            ps.executeUpdate();
                        }
                    }
                }
            }

            NodeList assets = (NodeList) xp.evaluate(
                "/FundsXML4/AssetMasterData/Asset", doc, XPathConstants.NODESET);
            for (int ai = 0; ai < assets.getLength(); ai++) {
                Node a = assets.item(ai);
                try (PreparedStatement ps = c.prepareStatement(
                        "INSERT INTO asset VALUES (?,?,?,?,?,?,?)")) {
                    ps.setString(1, docId);
                    ps.setString(2, t(a, "UniqueID"));
                    ps.setString(3, t(a, "Identifiers/ISIN"));
                    ps.setString(4, t(a, "Name"));
                    ps.setString(5, t(a, "AssetType"));
                    ps.setString(6, t(a, "Currency"));
                    ps.setString(7, t(a, "Country"));
                    ps.executeUpdate();
                }
            }
            c.commit();
            System.out.println("imported document_id: " + docId);
        }
    }
}

File: Database_Integration/javascript/import_fundsxml.mjs
// =============================================================================
// IMPORT  —  FundsXML file  ->  relational database  (Node.js, SQLite).
//
// Standalone, copy-me example of ONE direction (FundsXML -> DB). The reverse
// is a separate program, export_fundsxml.mjs. Over-commented as documentation.
//
// DB SCHEMA  ../ddl/schema.sql  (document -> fund -> portfolio -> position;
//   share_class per fund; asset document-scoped). Creates the schema in a
//   fresh DB then loads the file. The "db file" is sql.js's serialized image.
//
// RUN
//   cd Database_Integration/javascript && npm install
//   node import_fundsxml.mjs fx.db \
//     ../../FundsXML_Files/4.2.9/positions/Multi-Fund_Positions.xml
//   node export_fundsxml.mjs fx.db FUNDSXML_MULTI_1 out.xml
//
// DEPENDENCIES (pure-JS, no native build): sql.js (WASM SQLite),
//   @xmldom/xmldom + xpath (namespace-free DOM). Node 20 has no built-in
//   node:sqlite (added in 22.5), hence sql.js.
//
// FUNDSXML ASSUMPTIONS
//   * No XML namespace -> bare element names in XPath.
//   * Many <Fund>/<Portfolio>/<Position>: all iterated; 1-based *_seq columns
//     preserve order so the separate export reproduces the original document.
//   * Positions link to AssetMasterData by shared <UniqueID> -> `asset` is
//     document-scoped.
//
// SECURITY  @xmldom/xmldom never resolves external entities; we also reject a
//   DOCTYPE so a feed cannot declare entities (XXE).
// =============================================================================
import { readFileSync, writeFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import initSqlJs from "sql.js";
import { DOMParser } from "@xmldom/xmldom";
import xpath from "xpath";

const HERE = dirname(fileURLToPath(import.meta.url));
const DDL = resolve(HERE, "..", "ddl", "schema.sql");

const POSITION_KINDS = new Set(["Equity", "Bond", "ShareClass", "Warrant",
  "Certificate", "Option", "Future", "FXForward", "Swap", "Repo",
  "RealEstate", "CallMoney", "Account", "Generic"]);
const QTY_ELEM = { Equity: "Units", Warrant: "Units", Certificate: "Units",
  Bond: "Nominal", ShareClass: "Shares", Option: "Contracts",
  Future: "Contracts" };

const txt = (ctx, expr) => {            // first match's text or null
  const n = xpath.select1(expr, ctx);
  return n && n.textContent !== "" && n.textContent != null
    ? n.textContent : null;
};
const numOrNull = (s) => (s == null || s === "" ? null : Number(s));

function execScript(db, sql) {
  // Strip "--" line comments (no string literals in the DDL) then run.
  const clean = sql.split("\n").map((l) => {
    const i = l.indexOf("--");
    return i >= 0 ? l.slice(0, i) : l;
  }).join("\n");
  for (const stmt of clean.split(";")) if (stmt.trim()) db.run(stmt);
}

const [dbPath, xmlPath] = process.argv.slice(2);
if (!dbPath || !xmlPath) {
  console.error("usage: import_fundsxml.mjs <db> <fundsxml.xml>");
  process.exit(2);
}

const src = readFileSync(xmlPath, "utf8");
if (/<!DOCTYPE/i.test(src)) throw new Error("DOCTYPE not allowed (XXE)");
const doc = new DOMParser().parseFromString(src, "text/xml");

const SQL = await initSqlJs();
const db = new SQL.Database();
execScript(db, readFileSync(DDL, "utf8"));   // create the schema (fresh DB)

const cd = xpath.select1("/FundsXML4/ControlData", doc);
const docId = txt(cd, "UniqueDocumentID");

db.run("INSERT INTO document VALUES (?,?,?,?,?,?,?,?,?)", [
  docId, txt(cd, "DocumentGenerated"), txt(cd, "Version"),
  txt(cd, "ContentDate"), txt(cd, "DataOperation"),
  txt(cd, "DataSupplier/SystemCountry"), txt(cd, "DataSupplier/Short"),
  txt(cd, "DataSupplier/Name"), txt(cd, "DataSupplier/Type")]);

xpath.select("/FundsXML4/Funds/Fund", doc).forEach((fund, fi) => {
  const fundSeq = fi + 1;                     // 1-based document order
  const ccy = txt(fund, "Currency");
  const tav = xpath.select1(
    "FundDynamicData/TotalAssetValues/TotalAssetValue", fund);
  db.run("INSERT INTO fund VALUES (?,?,?,?,?,?,?,?)", [
    docId, fundSeq, txt(fund, "Identifiers/LEI"),
    txt(fund, "Names/OfficialName"), ccy, txt(fund, "SingleFundFlag"),
    txt(tav, "NavDate"),
    Number(txt(tav, `TotalNetAssetValue/Amount[@ccy='${ccy}']`))]);

  for (const sc of xpath.select("SingleFund/ShareClasses/ShareClass", fund)) {
    db.run("INSERT INTO share_class VALUES (?,?,?,?,?,?,?,?)", [
      docId, fundSeq, txt(sc, "Identifiers/ISIN"),
      txt(sc, "Names/OfficialName"), txt(sc, "Currency"),
      numOrNull(txt(sc, "Prices/Price/NavPrice")),
      numOrNull(txt(sc,
        `TotalAssetValues/TotalAssetValue/TotalNetAssetValue/Amount[@ccy='${ccy}']`)),
      numOrNull(txt(sc,
        "TotalAssetValues/TotalAssetValue/SharesOutstanding"))]);
  }

  xpath.select("FundDynamicData/Portfolios/Portfolio", fund)
    .forEach((port, pi) => {
      const portSeq = pi + 1;
      db.run("INSERT INTO portfolio VALUES (?,?,?,?)",
        [docId, fundSeq, portSeq, txt(port, "NavDate")]);
      xpath.select("Positions/Position", port).forEach((pos, qi) => {
        let kind = null;
        for (let ch = pos.firstChild; ch; ch = ch.nextSibling)
          if (ch.nodeType === 1 && POSITION_KINDS.has(ch.nodeName)) {
            kind = ch.nodeName; break;
          }
        const qty = kind && QTY_ELEM[kind]
          ? numOrNull(txt(pos, `${kind}/${QTY_ELEM[kind]}`)) : null;
        db.run("INSERT INTO position VALUES (?,?,?,?,?,?,?,?,?,?,?)", [
          docId, fundSeq, portSeq, qi + 1, txt(pos, "UniqueID"),
          txt(pos, "Identifiers/ISIN"), txt(pos, "Currency"),
          Number(txt(pos, `TotalValue/Amount[@ccy='${ccy}']`)),
          Number(txt(pos, "TotalPercentage")), kind, qty]);
      });
    });
});

for (const a of xpath.select("/FundsXML4/AssetMasterData/Asset", doc)) {
  db.run("INSERT INTO asset VALUES (?,?,?,?,?,?,?)", [
    docId, txt(a, "UniqueID"), txt(a, "Identifiers/ISIN"),
    txt(a, "Name"), txt(a, "AssetType"), txt(a, "Currency"),
    txt(a, "Country")]);
}

writeFileSync(dbPath, Buffer.from(db.export()));
console.log("imported document_id:", docId);
